Lots of -Wundef warnings

The preprocessor checks for the current compiler trigger a lot of `-Wundef` warnings with GCC, which is a bit noisy. Here is a small excerpt:

```cpp
[...]/libassert-src/include/assert.hpp:48:5: warning: "LIBASSERT_IS_CLANG" is not defined, evaluates to 0 [-Wundef]
   48 | #if LIBASSERT_IS_CLANG || LIBASSERT_IS_GCC
      |     ^~~~~~~~~~~~~~~~~~
[...]/libassert-src/include/assert.hpp:60:5: warning: "LIBASSERT_IS_MSVC" is not defined, evaluates to 0 [-Wundef]
   60 | #if LIBASSERT_IS_MSVC
      |     ^~~~~~~~~~~~~~~~~
[...]/libassert-src/include/assert.hpp:48:5: warning: "LIBASSERT_IS_CLANG" is not defined, evaluates to 0 [-Wundef]
   48 | #if LIBASSERT_IS_CLANG || LIBASSERT_IS_GCC
      |     ^~~~~~~~~~~~~~~~~~
[...]/libassert-src/include/assert.hpp:60:5: warning: "LIBASSERT_IS_MSVC" is not defined, evaluates to 0 [-Wundef]
   60 | #if LIBASSERT_IS_MSVC
      |     ^~~~~~~~~~~~~~~~~
[...]/libassert-src/include/assert.hpp:526:13: warning: "LIBASSERT_IS_CLANG" is not defined, evaluates to 0 [-Wundef]
  526 |         #if LIBASSERT_IS_CLANG
      |             ^~~~~~~~~~~~~~~~~~
[...]/libassert-src/include/assert.hpp:526:13: warning: "LIBASSERT_IS_CLANG" is not defined, evaluates to 0 [-Wundef]
  526 |         #if LIBASSERT_IS_CLANG
      |             ^~~~~~~~~~~~~~~~~~
[...]/libassert-src/include/assert.hpp:827:17: warning: "LIBASSERT_IS_MSVC" is not defined, evaluates to 0 [-Wundef]
  827 |             #if LIBASSERT_IS_MSVC
      |                 ^~~~~~~~~~~~~~~~~
```

To give an idea of the scale: I was trying to add optional libassert support for all assertions in cpp-sort and immediately got 4500 lines of those warnings.

--- a/Makefile
+++ b/Makefile
@@ -20,7 +20,7 @@ ifneq ($(COMPILER),msvc)
     # GCC / Clang
     CPP = $(COMPILER)
     LD = $(COMPILER)
-    override WFLAGS += -Wall -Wextra -Wvla -Wshadow -Werror
+    override WFLAGS += -Wall -Wextra -Wvla -Wshadow -Werror -Wundef
     override FLAGS += -std=$(STD) $(GFLAG) -Iinclude
     override LDFLAGS += -Wl,--whole-archive -Wl,--no-whole-archive
     ifeq ($(TARGET), debug)
@@ -60,7 +60,7 @@ else
     CPP = cl
     LD = link
     SHELL = powershell
-    override WFLAGS += /W4 /WX
+    override WFLAGS += /W4 /WX /wd4668
     override FLAGS += /std:$(STD) /EHsc
     override LDFLAGS += /WHOLEARCHIVE # /PDB /OPT:ICF /OPT:REF
     ifeq ($(TARGET), debug)

--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -24,7 +24,7 @@
  #pragma error "libassert requires C++17 or newer"
 #endif
 
-#if __cplusplus >= 202002L || _MSVC_LANG >= 202002L
+#if __cplusplus >= 202002L || (defined(_MSVC_LANG) && _MSVC_LANG >= 202002L)
  #include <compare>
 #endif
 
@@ -34,11 +34,18 @@
  "../third_party/magic_enum.hpp"
 #endif
 
+#define LIBASSERT_IS_CLANG 0
+#define LIBASSERT_IS_GCC 0
+#define LIBASSERT_IS_MSVC 0
+
 #if defined(__clang__)
+ #undef LIBASSERT_IS_CLANG
  #define LIBASSERT_IS_CLANG 1
 #elif defined(__GNUC__) || defined(__GNUG__)
+ #undef LIBASSERT_IS_GCC
  #define LIBASSERT_IS_GCC 1
 #elif defined(_MSC_VER)
+ #undef LIBASSERT_IS_MSVC
  #define LIBASSERT_IS_MSVC 1
  #include <iso646.h> // alternative operator tokens are standard but msvc requires the include or /permissive- or /Za
 #else

--- a/src/assert.cpp
+++ b/src/assert.cpp
@@ -19,7 +19,11 @@
 #include <unordered_map>
 #include <unordered_set>
 
+#define IS_WINDOWS 0
+#define IS_LINUX 0
+
 #if defined(_WIN32)
+ #undef IS_WINDOWS
  #define IS_WINDOWS 1
  #ifndef STDIN_FILENO
   #define STDIN_FILENO _fileno(stdin)
@@ -35,6 +39,7 @@
  #undef max
  #define USE_DBG_HELP_H
 #elif defined(__linux)
+ #undef IS_LINUX
  #define IS_LINUX 1
  #include <execinfo.h>
  #include <climits>

